runtime/shared_heap.c

```diff
@@ -785,6 +811,475 @@ void caml_verify_heap(caml_domain_state *domain) {
+        p += wh;
+      }
+
+      evac_pool = evac_pool->next;
+    }
+
+    /* A pool we can allocate in to must have empty space we can allocate
+      a new block in to else it would should have been moved to the full pools
+      list. */
+    CAMLassert(cur_pool->next_obj != NULL);
+  }
+
+  caml_global_barrier();
+
+  /* Second phase: at this point all live blocks in evacuated pools have been
+    moved and their old location's first field now points to their new location.
+    We now go through all pools again (including full ones this time)
+    and for each field we check what shared pool it is in and if it an
+    evacuating pool. This hopefully generates fewer cache misses than using
+    a bit in the header of moved blocks. If the pool is evacuating we find
+    the moved block's forwarding pointer and update accordingly. */
+
+  /* First we do roots (locals and finalisers) */
+  caml_do_roots(&compact_update_field, 0, NULL, Caml_state, 1);
+
+  /* Next, one domain does the global roots */
+  if( participants[0] == Caml_state ) {
+    caml_scan_global_roots(&compact_update_field, NULL);
+  }
+
+  /* Shared heap pools. */
+  for(int sz_class = 1; sz_class < NUM_SIZECLASSES; sz_class++) {
+    pool* cur_pool = heap->unswept_avail_pools[sz_class];
+
+    /* Iterate through pools updating the fields in each live block,
+      stop once we reach our first evacuating pool. */
+    while( cur_pool != NULL && cur_pool->evacuating == 0 ) {
+      value* p = (value*)((char*)cur_pool + POOL_HEADER_SZ);
+      value* end = (value*)cur_pool + POOL_WSIZE;
+      mlsize_t wh = wsize_sizeclass[sz_class];
+
+      while (p + wh <= end) {
+        if( *p != 0
+          && Has_status_val(Val_hp(p), caml_global_heap_state.UNMARKED) ) {
+          compact_update_block(p);
+        }
+          p += wh;
+      }
+
+      cur_pool = cur_pool->next;
+    }
+
+    /* do the same for full pools */
+    cur_pool = heap->unswept_full_pools[sz_class];
+
+    while( cur_pool != NULL ) {
+      CAMLassert( cur_pool->evacuating == 0 );
+      value* p = (value*)((char*)cur_pool + POOL_HEADER_SZ);
+      value* end = (value*)cur_pool + POOL_WSIZE;
+      mlsize_t wh = wsize_sizeclass[sz_class];
+
+      while (p + wh <= end) {
+        CAMLassert( *p != 0 );
+        if( Has_status_val(Val_hp(p), caml_global_heap_state.UNMARKED) ) {
+          compact_update_block(p);
+        }
+          p += wh;
+      }
+
+      cur_pool = cur_pool->next;
+    }
+  }
+
+  /* Large allocations */
+  for( large_alloc* la = heap->unswept_large;
+        la != NULL; la = la->next ) {
+    value* p = (value*)((char*)la + LARGE_ALLOC_HEADER_SZ);
+    if( Has_status_val(Val_hp(p), caml_global_heap_state.UNMARKED) ) {
+      compact_update_block(p);
+    }
+  }
+
+  /* Ephemerons */
+  struct caml_ephe_info* ephe_info = Caml_state->ephe_info;
+
+  if( ephe_info->todo ) {
+    compact_update_field(NULL, ephe_info->todo, &ephe_info->todo);
+    compact_update_ephe_list(ephe_info->todo);
```

Comment:
If `compact_update_ephe_list` took the address of the header instead, it could include the condition and the initial `compact_update_field` so this would just be
```
compact_update_ephe_list(&ephe_info->todo);
compact_update_ephe_list(&ephe_info->live);
```